### DISCUSSION TITLE
Investigate integration with Microsoft.Toolkit.HighPerformance

### DISCUSSION THREAD

**Discussion by Sergio0694** [SixLabors/ImageSharp#1392] [Ideas] [answered]
Hey @JimBobSquarePants and team, wanted to run an idea by you guys 😄
I noticed there's quite some overlap between lots of code in `ImageSharp` and two packages in the toolkit, specifically the base `Microsoft.Toolkit` and the new `Microsoft.Toolkit.HighPerformance` package. In particular:

- Virtually the entire `Guard` class is present in `Microsoft.Toolkit` package, possibly with even more APIs there. Related docs [here](https://docs.microsoft.com/dotnet/api/microsoft.toolkit.diagnostics.guard?view=win-comm-toolkit-dotnet-stable). The codegen in these APIs should even be slightly better than that of `ImageSharp`, in fact I was currently backporting some of the tweaks there to `ImageSharp` in https://github.com/SixLabors/SharedInfrastructure/pull/12.
- I've added a very simple API to wrap `Memory<byte>` to arbitrary pixel types in https://github.com/SixLabors/ImageSharp/pull/1314, to close [this feature request](https://github.com/SixLabors/ImageSharp/issues/1097)https://github.com/SixLabors/ImageSharp/issues/1097) in `ImageSharp`. I'm now adding a much better version of this API to the `Microsoft.Toolkit.HighPerformance` package ([here](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3520)), which is more flexible but also more efficient, as it skips a number of indirections that are present in the API currently in `ImageSharp`.
- I've also added a comprehensive set of `ThrowHelper` APIs to the `Microsoft.Toolkit` package ([here](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3346) and [here](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3455)) that we could use throughout `ImageSharp` to improve the codegen (I've noticed some inline `throw new` expressions all over the codebase).
- There's also a number of utility methods that I've added in the whole `Microsoft.Toolkit.HighPerformance` that might be useful to have as well in various `ImageSharp` internal methods, without having to reinvent the wheel every time and test all that.

I'm fully aware that `ImageSharp` is meant to be completely self-contained and with no dependencies, so I know that just referencing the two NuGet packages mentioned above would probably not be feasible (though note that both those packages mentioned above are fully self-contained as well and only use standard `System.*` assemblies and nothing else).

I was wondering, would this be something you might be interested in if it came with no extra package dependencies, for instance via a git submodule that directly points to the source of the two packages in the Toolkit repo ([here](https://aka.ms/wct))? I'd be more than happy to help out with the integration and with some refactoring in `ImageSharp` to switch the APIs to those from the toolkit, of course 😊

---

### Other examples

Was going through some bits of code in `ImageSharp` to find other examples of types from the `HighPerformance` package that could be useful, and I find some instances where I think the new set of 2D memory types I've been working on would be pretty useful to make the code less error prone and much easier to read. The types are `Memory2D<T>`, `Span2D<T>` and others (full list of APIs and related PR [here](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3353)).

I was these conversion methods from `System.Drawing` to `ImageSharp` images ([source](https://github.com/SixLabors/ImageSharp/blob/b06cb32b7114961fd5473f7645d38f8fee04ec64/tests/ImageSharp.Tests/TestUtilities/ReferenceCodecs/SystemDrawingBridge.cs#L85)) that could be rewritten using `Span2D<T>` to be clearer and less verbose. Here's an example of that method linked above, rewritten using these new APIs:

```csharp
internal static unsafe Image<TPixel> From24bppRgbSystemDrawingBitmap2<TPixel>(Bitmap bmp)
    where TPixel : unmanaged, IPixel<TPixel>
{
    int w = bmp.Width;
    int h = bmp.Height;

    var fullRect = new Rectangle(0, 0, w, h);

    BitmapData data = bmp.LockBits(fullRect, ImageLockMode.ReadWrite, bmp.PixelFormat);
    
    try
    {
        var image = new Image<TPixel>(w, h);

        Configuration configuration = image.GetConfiguration();

        Span2D<Bgr24> source = new Span2D<Bgr24>((void*)data.Scan0, h, w, 0);

        for (int y = 0; y < h; y++)
        {
            PixelOperations<TPixel>.Instance.FromBgr24(configuration, source.GetRowSpan(y), image.GetPixelRowSpan(y));
        }

        return image;
    }
    finally
    {
        bmp.UnlockBits(data);
    }
}
```

You can see how the `Span2D<T>` type automatically takes care of calculating the right indexing into the source memory area. The same type can also track an internal pitch, which means it can support discontiguous memory regions as well, as in, memory regions that are cropped within a larger memory area (the source area is contiguous). I think there could be other use cases for these APIs as well in some places within the codebase, just like using the type here made the code much less verbose due to not having to manually calculate the right offset into each row manually. It basically acts as an abstraction API in the same way that `Memory<T>` and `Span<T>` work, but in 2D.

Let me know what you think! 🚀

**Comment by antonfirsov** [comment, marked as answer]
You probably won't like my answer.

We have our own versions of `Memory2D` (`Buffer2D`) and `Span2D` (`Buffer2DRegion`) which are already integrated, and work efficiently with other ImageSharp memory management primitives. In theory we could unify these middleware components, which would be an interesting technical challenge, but doesn't bring visible customer benefits for ImageSharp.

Regarding `Guard`, `ThrowHelper` and similar: I believe these are very thin primitives, where a potential benefit of a dependency doesn't justify the cost of taking a dependency. I believe we get the most productivity and flexibility if the source of these primitives is self-contained & maintained by SixLabors.

In short, I'd see such an experiment a distraction from our top goals, with little or no benefits. Our time is **super limited** these days, and we need it to be 100% focused on our backlog:
https://github.com/SixLabors/ImageSharp/issues
https://github.com/SixLabors/ImageSharp.Drawing/issues

**Reply by JimBobSquarePants** [reply]
I'm afraid have to agree with @antonfirsov here. 

There's some really, really cool stuff in the library (some of which I've already stolen for my ZlibStream implementation) but we cannot dedicate any time to looking into this. Even if we could I don't think it would be possible since the minimum netstandard target for the library is 1.4 (we're 1.3)

**Reply by Sergio0694** [reply]
Hey @antonfirsov and @JimBobSquarePants - fair enough, I just thought I'd ask 😄

Thank you for taking the time to read that and to reply, appreciate it. I understand that taking a dependency was out of the question and that's why I suggested maybe using a submodule or investigating other options, but even then I do understand how this would take some time to do the actual refactoring, that makes perfect sense.

> There's some really, really cool stuff in the library (some of which I've already stolen for my ZlibStream implementation)

@JimBobSquarePants Well first of all I'm glad to hear that, thanks! About what you took for your ZlibStream implementation, could you elaborate on that when you have some time? Part of this issue was also just for me to investigate examples of where the package could be helpful in other libraries, so I'd still be interested in hearing about how you could reuse bits of that in your other projects, even if you didn't reference the package directly. Also as a way for me to understand what bits of the package are more useful to other devs 😄

Thanks again for your time guys, and keep up the awesome work on `ImageSharp`! 🍻

**Reply by JimBobSquarePants** [reply]
@Sergio0694 Yoink!

https://github.com/SixLabors/ZlibStream/blob/21c608cf57c038e48f97a949a56de6b11ffc2c45/src/ZlibStream/ArrayExtensions.cs#L24

I use to to allow fast access of tree data for static huffman trees. When I saw you publish it that allowed me to do a huge refactor.
https://github.com/SixLabors/ZlibStream/blob/21c608cf57c038e48f97a949a56de6b11ffc2c45/src/ZlibStream/Trees.cs#L190-L194

I'll be able to encode PNGs 2x faster soon! 😄

**Reply by Sergio0694** [reply]
Ha, that's pretty neat, thanks for sharing, and glad to see that was useful for you! 😊

**Reply by antonfirsov** [reply]
Maaan .. you really like to be on adrenaline! 😆 What if a future change in runtime breaks this?

`DistCode` could be `ReadOnlySpan<byte>` thx to dotnet/roslyn#24621:
```C#
private static readonly ReadOnlySpan<byte> DistCode => new byte[] { ... }
```

And then fast unsafe indexing can be achieved with `MemoryMarshal.GetReference` + `Unsafe.Add`.

**Reply by Sergio0694** [reply]
Oh yeah, didn't notice that lookup table was entirely made up of values that fit in a `byte`, in that case you should definitely go with a `ReadOnlySpan<byte>` mapping to the .text segment, that's safer and faster too across all runtimes 👍

> What if a future change in runtime breaks this?

@antonfirsov On .NET 5 that method should be updated to a call to `MemoryMarshal.GetArrayDataReference` 🚀
That hack is only meant to be used on .NET Core 2.1 and 3.1, so it's guaranteed to work there.

**Reply by antonfirsov** [reply]
There is no safe way to guarantee backwards compatibility here. A future .NET 13.0 app can take a reference to a `netcoreapp3.1` dll which is compiled against this stuff, and then probably crash.

In my previous job, I've seen a case where a poorly supported internal library based a critical feature on .NET Framework runtime internals in a similar manner. After a .NET Framework upgrade, one our components started to crash in a magical way. My collogues spent a weekend with desperate debugging + disassembling to figure out what went wrong.

**Reply by Sergio0694** [reply]
Ok, well that's pretty terrifying 🤣

I mean, good thing those methods are all called `Dangerous___`, and I for one I'm eager to eventually remove all these hacks relying on internals. In fact the current version of the `HighPerformance` package has a different implementation on .NET 5 there which is fully portable and doesn't rely on internals at all (did so as the .NET 5 target could run on Mono LLVM as well). Hopefully people won't have to go through all that debugging work in the future, or rather, these things should hopefully not happen 😄

**Reply by JimBobSquarePants** [reply]
That’s one of only two places I’m using it. The other place is an array of integers so can’t  use ROS

**Reply by JimBobSquarePants** [reply]
For NET 5 we can use the new MemoryMarshal APIs to avoid all the hacking as mentioned above.

**Reply by JimBobSquarePants** [reply]
Actually, just realized I was looking at the wrong table! Can use MM and ROS for both. Hurrah!
